Qiskit Services through REST APIs under `docs/run` section (#1492)
Qiskit/documentation: docs/run/rest-api-primitives.mdx

---
title: Primitives using REST API
description: How to use the Sampler and Estimator primitives with Qiskit Runtime REST API.
---

# Primitives using REST API

The steps in this topic describe how to run primitives workloads with REST API, explore the options you can use to configure them, and invoke them in any program of your choice.

<Admonition type="note">
    This documentation utilizes the `requests` module of python to demonstrate the Qiskit Runtime REST API. However, this workflow can be executed using any language/framework that supports working with REST APIs. The official documentation for the API Reference can be found [here](https://docs.quantum.ibm.com/api/runtime/tags/jobs).
</Admonition>

<span id="start-estimator"></span>
## Estimator Primitive with REST API

### 1. Initialize the account

> Because Qiskit Runtime `Estimator` is a managed service, you first need to initialize your account. You can then select the simulator or real system you want to use to calculate the expectation value. You can then select the IBM Quantum Compute resource of your choice to run your calculations on. 

Reviewer: ```suggestion
Because Qiskit Runtime `Estimator` is a managed service, you first need to initialize your account. You can then select the device you want to use to calculate the expectation value. You can then select the IBM Quantum Compute resource of your choice to run your calculations on. 
```
we don't provide access to cloud simulators anymore (at least for the majority of users). Also we are supposed to be moving away from the term "system" unless we are specifically referring to System Two

Details on how to initialize your account, view available backends and invalidate tokens can be found in this [topic](/start/setup-channel#access-ibm-quantum-systems-using-rest-api).

### 2. Create a QASM circuit

You need at least one circuit as the input to the `Estimator` primitive.

> * Define a QASM quantum circuit e.g. 

Reviewer: ```suggestion
Define a QASM quantum circuit e.g. 
```
This was rendering with a bullet point, which I assumed was not intentional? This change should fix that

Author: Thanks for catching that.

```python
qasm_string='''
OPENQASM 3;
include "stdgates.inc";
qreg q[2];
creg c[2];
x q[0];
cx q[0], q[1];
c[0] = measure q[0]; 
c[1] = measure q[1];
'''  
```

The code snippets given below assume that the `qasm_string` has been transpiled to a new string `resulting_qasm`. Detailed instructions on how to use the Cloud Transpiler API can be found in this [topic](/transpile/rest-api-transpilation).

### 3. Run the Quantum Circuit using the EstimatorV2 API

<Admonition type="note">

>     The jobs below use [Qiskit Runtime V2 primitives](https://docs.quantum.ibm.com/api/migration-guides/v2-primitives). Both SamplerV2 and EstimatorV2 take one or more primitive unified blocs (PUBs) as the input. Each PUB is a tuple that contains one circuit and the data broadcasted to that circuit, which can be multiple observables and parameters. Each PUB returns a result.

Reviewer: ```suggestion
    The jobs below use [Qiskit Runtime V2 primitives](https://docs.quantum.ibm.com/api/migration-guides/v2-primitives). Both `SamplerV2` and `EstimatorV2` take one or more primitive unified blocs (PUBs) as the input. Each PUB is a tuple that contains one circuit and the data broadcasted to that circuit, which can be multiple observables and parameters. Each PUB returns a result.
```

else:
    print(f"Error: {response.status_code}")
```

### 4. Check Job Status and Get Results

Next, pass the `job_id` to the API:

```python
response_status_singlejob= requests.get(url+'/'+job_id, headers=headers)
response_status_singlejob.json().get('state')
```
Output

```text
>>> Job ID: 58223448-5100-4dec-a47a-942fb30edcad
>>> Job Status: JobStatus.RUNNING
```

Get job results:
```python
response_result= requests.get(url+'/'+job_id+'/results', headers=headers)

res_dict=response_result.json()

estimator_result=res_dict['results']
print(estimator_result)
```
Output
```text
[{'data': {'evs': 0.7428980350102542, 'stds': 0.029884014518789213, 'ensemble_standard_error': 0.03261147170624149}, 'metadata': {'shots': 10016, 'target_precision': 0.01, 'circuit_metadata': {}, 'resilience': {}, 'num_randomizations': 32}}]
```

### 5. Working with Runtime Options

Error mitigation techniques allow users to mitigate circuit errors by modeling the device noise at the time of execution. This typically results in quantum pre-processing overhead related to model training and classical post-processing overhead to mitigate errors in the raw results by using the generated model.

> The error mitigation techniques built in to primitives are advanced resilience options. To specify these options, use the resilience_level option when submitting your job.

Reviewer: ```suggestion
The error mitigation techniques built in to primitives are advanced resilience options. To specify these options, use the `resilience_level` option when submitting your job.
```

> You will see examples of three such error mitigation / suppression techniques when used with the Estimator Primitive. The examples given below demonstrate the default options for error mitigation / suppression techniques but more options for [Dynamical decoupling](https://github.com/Qiskit/qiskit-ibm-runtime/blob/main/qiskit_ibm_runtime/options/dynamical_decoupling_options.py), [Twirling](https://github.com/Qiskit/qiskit-ibm-runtime/blob/main/qiskit_ibm_runtime/options/twirling_options.py) and [TREX + ZNE](https://github.com/Qiskit/qiskit-ibm-runtime/blob/main/qiskit_ibm_runtime/options/resilience_options.py) can be further explored. 

Reviewer: Instead of all these links I think it would be good to link to the new error mitigation techniques page that we just released in docs: https://docs.quantum.ibm.com/run/error-mitigation-explanation

Author: Wow, this is quite helpful. I hadn't come across this page before.

                "enable_gates": True, 
                "enable_measure": True,
                "num_randomizations": "auto",
                "shots_per_randomization": "auto",
                "strategy": "active-accum",
                },
        },
    }
}

response = requests.post(url, headers=headers, json=job_input)

if response.status_code == 200:
    job_id = response.json().get('id')
    print("Job created:",response.text)
else:
    print(f"Error: {response.status_code}")
```

  </TabItem>
</Tabs>

<span id="start-sampler"></span>
## Sampler Primitive with REST API

### 1. Initialize the account

Because Qiskit Runtime `Sampler` is a managed service, you first need to initialize your account. You can then select the simulator or real system you want to use to calculate the expectation value. You can then select the IBM Quantum Compute resource of your choice to run your calculations on. 

Details on how to initialize your account, view available backends and invalidate tokens can be found in this [topic](/start/setup-channel#access-ibm-quantum-systems-using-rest-api).

### 2. Create a QASM circuit

You need at least one circuit as the input to the `Sampler` primitive.

> * Define a QASM quantum circuit e.g. 

Reviewer: ```suggestion
Define a QASM quantum circuit e.g. 
```
same as above, this is rendering as a weird bullet point, this should fix
